Text index: Add bloom filter into FST segments

ahmadov opened the pull request:
Add bloom filter into each FST segment to reduce the number of segments to read. Bloom filters are created from the first 3 characters (bytes) of each token. Bloom filters are stored in a separate `*.gin_bflt` file.

Also, delay initialization of term dictionary (FST) data until it's needed.

`BloomFilterHash::calculationBestPractices` computes the bloom filter size and number of hash functions for the given probability. Max probability of conflict is set to 0.001.

### Changelog category (leave one):
- Not for changelog (changelog entry is not required)

ahmadov committed:
Add bloom filter into FST segments

With bloom filter, we are expecting to reduce number of segments for
reading. Also, delay initialization of term dictionary (FST) data until
it's needed.

ahmadov committed:
Adapt FST dump tree tool to handle bloom filters

ahmadov committed:
Compute filter size and hashes for lower probability of conflict

BloomFilterHash::calculationBestPractices computes the bloom filter size
and number of hash functions for the given probability.

Max probability of conflict is set to 0.001.

ahmadov committed:
Use first and last m (n/2) characters instead of first n characters

ahmadov committed:
Merge remote-tracking branch 'origin/master' into ahmadov/fst-seg-bf

ahmadov committed:
Store bloom filters in a separate `.gin_filter` file

ahmadov committed:
Merge remote-tracking branch 'origin/master' into ahmadov/fst-seg-bf

ahmadov committed:
small fix after merge origin/master

ahmadov committed:
fix checksum for text index bloom filter file

ahmadov committed:
Merge remote-tracking branch 'origin/master' into ahmadov/fst-seg-bf

rschu1ze commented on src/Storages/MergeTree/GinIndexStore.cpp:
```diff
@@ -350,16 +418,21 @@ void GinIndexStore::writeSegment()
     TokenPostingsBuilderPairs token_postings_list_pairs;
     token_postings_list_pairs.reserve(current_postings.size());
 
+    GinSegmentDictionaryBloomFilter bloom_filter(0.001, GIN_INDEX_BLOOM_FILTER_DEFAULT_MAX_TOKEN_SIZE);
```
0.001 should be a constant

Ergus reviewed (commented):
LGTM

rschu1ze commented on src/Storages/MergeTree/GinIndexStore.cpp:
```diff
@@ -154,6 +155,65 @@ GinIndexPostingsListPtr GinIndexPostingsBuilder::deserialize(ReadBuffer & buffer
+    }
+    else
+    {
+        bloom_filter->add(token, first_characters);
```
Each `.add` on `bloom_filter` hashes the input and sets some bits in the bloom filter. If we insert the prefix and the postfix of each token independently (two `.add` calls), then is as if two tokens are inserted. I would think we either need to size the bloom filter differently (in the `GinSegmentDictionaryBloomFilter` ctor, essentially double the size to sustain the target false positive rate), or we need to concat the prefix and the suffix and call `.add` just once. Or we ignore the suffix completely but that is also not great.

EDIT: What if we hash the whole token? Thinking about it, I no longer understand the reasons for hashing prefix/suffix in the first place. Was it to save some time on computing hashes? These will likely be negligible in the grand scheme of things.

EDIT 2: I think we has only the prefixes/suffixes to reduce the overall cardinality, which leads to smaller bloom filters. Perhaps it makes sense to add the whole token in a first version. If the bloom filter is surprisingly large for a realistic dataset, we can optimize this case in a later version.

rschu1ze commented on src/Storages/MergeTree/GinIndexStore.cpp:
```diff
@@ -446,30 +524,53 @@ void GinIndexStoreDeserializer::initFileStreams()
 
-    metadata_file_stream->readStrict(reinterpret_cast<char *>(segments.data()), num_segments * sizeof(GinIndexSegment));
-    for (UInt32 i = 0; i < num_segments; ++i)
+    /// TODO(ahmadov): clean up versions and V1 handling
```
Let's keep our lifes simple for now (at least in the final version of this PR) and not add a new version. We are free to do that in the next 4 weeks.

ahmadov replied:
actually, I'm planning to remove V2 after this PR

Ergus commented:
I know it is tricky, but could we remember/plan to create a benchmark test for this? This adds an extra level of complexity with the intention to improve performance.

But we have not measured that benefit. AFAIS, the cost of reading index segments could be relatively small. But I would put more attention in how are we reading those segments.

As a no-fan of probabilistic data structures  I would like to have some sort of justification that support using them.

ahmadov committed:
Merge remote-tracking branch 'origin/master' into ahmadov/fst-seg-bf

ahmadov commented:
> I know it is tricky, but could we remember/plan to create a benchmark test for this? This adds an extra level of complexity with the intention to improve performance.
> 
> But we have not measured that benefit. AFAIS, the cost of reading index segments could be relatively small. But I would put more attention in how are we reading those segments.
> 
> As a no-fan of probabilistic data structures I would like to have some sort of justification that support using them.

exactly for the same reason, I added this performance test before opening this PR: https://github.com/ClickHouse/ClickHouse/blob/f5bdb49/tests/performance/text_index.xml

We don't see its effect on performance reports because we need some time to build a text index. There is a way to trigger it `OPTIMIZE TABLE hits_10m_single FINAL` but this operation is costly and takes some time to build, which we probably don't want to do it performance tests.

ahmadov commented on src/Storages/MergeTree/GinIndexStore.cpp:
```diff
@@ -154,6 +155,65 @@ GinIndexPostingsListPtr GinIndexPostingsBuilder::deserialize(ReadBuffer & buffer
+    last_characters = max_token_size_ >= 14 ? 7 : (max_token_size_ - first_characters);
+}
+
+void GinSegmentDictionaryBloomFilter::add(const char * token, UInt64 size)
```
the reason was getting the first and last characters easily for adding them into bloom filter. Since we don't need the first and last characters (see [comment)](https://github.com/ClickHouse/ClickHouse/pull/83485#discussion_r2213986245). We can now use `std::string_view`

ahmadov committed:
improve impl

ahmadov committed:
fix typo

ahmadov commented on src/Storages/MergeTree/GinIndexStore.h:
```diff
@@ -103,9 +137,15 @@ struct GinSegmentDictionary
     /// (Minimized) Finite State Transducer, which can be viewed as a map of <term, offset>, where offset is the
     /// offset to the term's posting list in postings list file
-    FST::FiniteStateTransducer offsets;
+    std::unique_ptr<FST::FiniteStateTransducer> fst;
```
FST: we delay the initialization of FST and I use `nullptr` to check if FST is initialized or not.

Bloom filter: V1 does not support, therefore it's nullptr for it - will remove V1 soon then we don't need unique_ptr anymore.

ahmadov committed:
naming changes

ahmadov committed:
Fix bloom filter size

Use a sample (10%) of data to compute an estimated unique count.

ahmadov commented on src/Storages/MergeTree/GinIndexStore.cpp:
```diff
@@ -350,16 +403,22 @@ void GinIndexStore::writeSegment()
     TokenPostingsBuilderPairs token_postings_list_pairs;
     token_postings_list_pairs.reserve(current_postings.size());
 
+    const auto [bits_per_rows, hashes] = BloomFilterHash::calculationBestPractices(GIN_INDEX_BLOOM_FILTER_DEFAULT_FALSE_POSITIVE_RATE);
```
@rschu1ze, thanks for the hint! I pushed a fix with 073ddec7f849add413f2e471c4019b87dd503a3e. To keep the overhead lower, I used a sample (10%) of input data to estimate a ratio of unique count.

ahmadov committed:
fix build

rschu1ze committed:
Merge remote-tracking branch 'ClickHouse/master' into ahmadov/fst-seg-bf

rschu1ze committed:
Make sampling thresholds configurable

rschu1ze commented on src/Storages/MergeTree/GinIndexStore.h:
```diff
@@ -157,6 +205,11 @@ class GinIndexStore
     UInt32 getCurrentSegmentID() const { return current_segment.segment_id; }
 
+    /// Set the unique value count, needed to size the bloom filter.
+    /// TODO: The unique row count is actually estimated per index granule but the bloom filter exists per segment.
```
Thinking about it, the sizing is better now (thanks), but I think we still need these two enhancements to make it perfect (--> fine to keep these TODOs around for now).

rschu1ze committed:
Merge remote-tracking branch 'ClickHouse/master' into ahmadov/fst-seg-bf

ahmadov committed:
Merge remote-tracking branch 'origin/master' into ahmadov/fst-seg-bf

ahmadov committed:
Fix build

Previously, the query context from the `MergeTreeIndexGin` was
forwarded during the index condition creation, but it was never
initialized, so it was `nullptr`.

Forward query context set by the caller of
`MergeTreeIndexGin::createIndexCondition`.

ahmadov committed:
Merge remote-tracking branch 'origin/master' into ahmadov/fst-seg-bf

rschu1ze committed:
Merge remote-tracking branch 'ClickHouse/master' into ahmadov/fst-seg-bf

rschu1ze committed:
Fix setting changes history

rschu1ze commented:
Looks we are still having a [test problem](https://s3.amazonaws.com/clickhouse-test-reports/json.html?PR=83485&sha=latest&name_0=PR&name_1=Stateless+tests+%28amd_binary%2C+ParallelReplicas%2C+s3+storage%2C+parallel%29) :-(

```
2025-07-25 11:21:24 Reason: return code:  105
2025-07-25 11:21:24 [c2d38eba7ed3] 2025.07.25 12:21:24.146303 [ 1055278 ] {fd0c751f-da05-46fd-a9cd-b7c734c1be6f} <Error> executeQuery: Code: 361. DB::Exception: Seek position is out of bounds. Offset: -1, Max: 0. (SEEK_POSITION_OUT_OF_BOUND) (version 25.8.1.1) (from [::1]:33250) (comment: 02346_text_index_queries.sql) (query 24, line 78) (in query: SELECT `__table1`.`k` AS `k`, `__table1`.`s` AS `s` FROM `test_m7lg3zis`.`tab_x` AS `__table1` WHERE hasToken(`__table1`.`s`, 'Alick') ORDER BY `__table1`.`k` ASC), Stack trace (when copying this message, always include the lines below):
2025-07-25 11:21:24 
2025-07-25 11:21:24 0. ./contrib/llvm-project/libcxx/include/__exception/exception.h:113: Poco::Exception::Exception(String const&, int) @ 0x000000001eb63432
2025-07-25 11:21:24 1. ./ci/tmp/build/./src/Common/Exception.cpp:115: DB::Exception::Exception(DB::Exception::MessageMasked&&, int, bool) @ 0x000000000f7eafdb
2025-07-25 11:21:24 2. DB::Exception::Exception(PreformattedMessage&&, int) @ 0x0000000008ded02c
2025-07-25 11:21:24 3. DB::Exception::Exception<long&, String>(int, FormatStringHelperImpl<std::type_identity<long&>::type, std::type_identity<String>::type>, long&, String&&) @ 0x000000000d56e86b
2025-07-25 11:21:24 4. ./ci/tmp/build/./src/IO/ReadBufferFromMemory.cpp:30: DB::ReadBufferFromMemoryHelper<DB::ReadBufferFromMemory>::seekImpl(long, int) @ 0x000000000f8cbc23
2025-07-25 11:21:24 5. ./src/IO/ReadBufferFromMemory.h:36: DB::FST::FiniteStateTransducer::getOutput(std::basic_string_view<char, std::char_traits<char>>) @ 0x000000000fab927a
2025-07-25 11:21:24 6. ./ci/tmp/build/./src/Storages/MergeTree/GinIndexStore.cpp:674: DB::GinIndexStoreDeserializer::readSegmentedPostingsLists(String const&) @ 0x0000000018471682
2025-07-25 11:21:24 7. ./ci/tmp/build/./src/Storages/MergeTree/GinIndexStore.cpp:697: DB::GinIndexStoreDeserializer::createPostingsCacheFromTerms(std::vector<String, std::allocator<String>> const&) @ 0x00000000184718e2
2025-07-25 11:21:24 8. ./ci/tmp/build/./src/Interpreters/GinFilter.cpp:190: DB::GinFilter::contains(DB::GinFilter const&, DB::PostingsCacheForStore&, DB::GinSearchMode) const @ 0x00000000186caafd
2025-07-25 11:21:24 9. ./ci/tmp/build/./src/Storages/MergeTree/MergeTreeIndexGin.cpp:290: DB::MergeTreeIndexConditionGin::mayBeTrueOnGranuleInPart(std::shared_ptr<DB::IMergeTreeIndexGranule>, DB::PostingsCacheForStore&) const @ 0x00000000186bcdd3
2025-07-25 11:21:24 10. ./ci/tmp/build/./src/Storages/MergeTree/MergeTreeDataSelectExecutor.cpp:1802: DB::MergeTreeDataSelectExecutor::filterMarksUsingIndex(std::shared_ptr<DB::IMergeTreeIndex const>, std::shared_ptr<DB::IMergeTreeIndexCondition>, std::shared_ptr<DB::IMergeTreeDataPart const>, DB::MarkRanges const&, DB::RangesInDataPartReadHints const&, DB::Settings const&, DB::MergeTreeReaderSettings const&, DB::MarkCache*, DB::UncompressedCache*, DB::VectorSimilarityIndexCache*, std::shared_ptr<Poco::Logger>) @ 0x000000001869bf24
2025-07-25 11:21:24 11. ./ci/tmp/build/./src/Storages/MergeTree/MergeTreeDataSelectExecutor.cpp:807: DB::MergeTreeDataSelectExecutor::filterPartsByPrimaryKeyAndSkipIndexes(DB::RangesInDataParts, std::shared_ptr<DB::StorageInMemoryMetadata const>, std::shared_ptr<DB::Context const> const&, DB::KeyCondition const&, std::optional<DB::KeyCondition> const&, std::optional<DB::KeyCondition> const&, DB::UsefulSkipIndexes const&, DB::MergeTreeReaderSettings const&, std::shared_ptr<Poco::Logger>, unsigned long, std::vector<DB::ReadFromMergeTree::IndexStat, std::allocator<DB::ReadFromMergeTree::IndexStat>>&, bool, bool, bool)::$_1::operator()(unsigned long) const @ 0x0000000018696fce
2025-07-25 11:21:24 12. ./ci/tmp/build/./src/Storages/MergeTree/MergeTreeDataSelectExecutor.cpp:856: DB::MergeTreeDataSelectExecutor::filterPartsByPrimaryKeyAndSkipIndexes(DB::RangesInDataParts, std::shared_ptr<DB::StorageInMemoryMetadata const>, std::shared_ptr<DB::Context const> const&, DB::KeyCondition const&, std::optional<DB::KeyCondition> const&, std::optional<DB::KeyCondition> const&, DB::UsefulSkipIndexes const&, DB::MergeTreeReaderSettings const&, std::shared_ptr<Poco::Logger>, unsigned long, std::vector<DB::ReadFromMergeTree::IndexStat, std::allocator<DB::ReadFromMergeTree::IndexStat>>&, bool, bool, bool) @ 0x0000000018693d3b
2025-07-25 11:21:24 13. ./ci/tmp/build/./src/Processors/QueryPlan/ReadFromMergeTree.cpp:2017: DB::ReadFromMergeTree::selectRangesToRead(DB::RangesInDataParts, std::shared_ptr<DB::MergeTreeData::IMutationsSnapshot const>, std::optional<DB::VectorSearchParameters> const&, std::shared_ptr<DB::StorageInMemoryMetadata const> const&, DB::SelectQueryInfo const&, std::shared_ptr<DB::Context const>, unsigned long, std::shared_ptr<std::unordered_map<String, long, std::hash<String>, std::equal_to<String>, std::allocator<std::pair<String const, long>>> const>, DB::MergeTreeData const&, std::vector<String, std::allocator<String>> const&, std::shared_ptr<Poco::Logger>, std::optional<DB::ReadFromMergeTree::Indexes>&, bool) @ 0x000000001921afbc
2025-07-25 11:21:24 14. ./ci/tmp/build/./src/Processors/QueryPlan/ReadFromMergeTree.cpp:1687: DB::ReadFromMergeTree::selectRangesToRead(bool) const @ 0x000000001921959d
2025-07-25 11:21:24 15. ./ci/tmp/build/./src/Processors/QueryPlan/ReadFromMergeTree.cpp:2282: DB::ReadFromMergeTree::initializePipeline(DB::QueryPipelineBuilder&, DB::BuildQueryPipelineSettings const&) @ 0x000000001922327e
2025-07-25 11:21:24 16. ./ci/tmp/build/./src/Processors/QueryPlan/ISourceStep.cpp:20: DB::ISourceStep::updatePipeline(std::vector<std::unique_ptr<DB::QueryPipelineBuilder, std::default_delete<DB::QueryPipelineBuilder>>, std::allocator<std::unique_ptr<DB::QueryPipelineBuilder, std::default_delete<DB::QueryPipelineBuilder>>>>, DB::BuildQueryPipelineSettings const&) @ 0x00000000191b2e4c
2025-07-25 11:21:24 17. ./ci/tmp/build/./src/Processors/QueryPlan/QueryPlan.cpp:202: DB::QueryPlan::buildQueryPipeline(DB::QueryPlanOptimizationSettings const&, DB::BuildQueryPipelineSettings const&, bool) @ 0x00000000191e4144
2025-07-25 11:21:24 18. ./ci/tmp/build/./src/Interpreters/InterpreterSelectQueryAnalyzer.cpp:289: DB::InterpreterSelectQueryAnalyzer::buildQueryPipeline() @ 0x00000000157ab917
2025-07-25 11:21:24 19. ./ci/tmp/build/./src/Interpreters/InterpreterSelectQueryAnalyzer.cpp:256: DB::InterpreterSelectQueryAnalyzer::execute() @ 0x00000000157ab65c
2025-07-25 11:21:24 20. ./ci/tmp/build/./src/Interpreters/executeQuery.cpp:1561: DB::executeQueryImpl(char const*, char const*, std::shared_ptr<DB::Context>, DB::QueryFlags, DB::QueryProcessingStage::Enum, DB::ReadBuffer*, std::shared_ptr<DB::IAST>&, std::shared_ptr<DB::ImplicitTransactionControlExecutor>) @ 0x0000000015acd78d
2025-07-25 11:21:24 21. ./ci/tmp/build/./src/Interpreters/executeQuery.cpp:1769: DB::executeQuery(String const&, std::shared_ptr<DB::Context>, DB::QueryFlags, DB::QueryProcessingStage::Enum) @ 0x0000000015ac79da
```

ahmadov commented:
> Looks we are still having a [test problem](https://s3.amazonaws.com/clickhouse-test-reports/json.html?PR=83485&sha=latest&name_0=PR&name_1=Stateless+tests+%28amd_binary%2C+ParallelReplicas%2C+s3+storage%2C+parallel%29) :-(
> 
> ```
> 2025-07-25 11:21:24 Reason: return code:  105
> 2025-07-25 11:21:24 [c2d38eba7ed3] 2025.07.25 12:21:24.146303 [ 1055278 ] {fd0c751f-da05-46fd-a9cd-b7c734c1be6f} <Error> executeQuery: Code: 361. DB::Exception: Seek position is out of bounds. Offset: -1, Max: 0. (SEEK_POSITION_OUT_OF_BOUND) (version 25.8.1.1) (from [::1]:33250) (comment: 02346_text_index_queries.sql) (query 24, line 78) (in query: SELECT `__table1`.`k` AS `k`, `__table1`.`s` AS `s` FROM `test_m7lg3zis`.`tab_x` AS `__table1` WHERE hasToken(`__table1`.`s`, 'Alick') ORDER BY `__table1`.`k` ASC), Stack trace (when copying this message, always include the lines below):
> 2025-07-25 11:21:24 
> 2025-07-25 11:21:24 0. ./contrib/llvm-project/libcxx/include/__exception/exception.h:113: Poco::Exception::Exception(String const&, int) @ 0x000000001eb63432
> 2025-07-25 11:21:24 1. ./ci/tmp/build/./src/Common/Exception.cpp:115: DB::Exception::Exception(DB::Exception::MessageMasked&&, int, bool) @ 0x000000000f7eafdb
> 2025-07-25 11:21:24 2. DB::Exception::Exception(PreformattedMessage&&, int) @ 0x0000000008ded02c
> 2025-07-25 11:21:24 3. DB::Exception::Exception<long&, String>(int, FormatStringHelperImpl<std::type_identity<long&>::type, std::type_identity<String>::type>, long&, String&&) @ 0x000000000d56e86b
> 2025-07-25 11:21:24 4. ./ci/tmp/build/./src/IO/ReadBufferFromMemory.cpp:30: DB::ReadBufferFromMemoryHelper<DB::ReadBufferFromMemory>::seekImpl(long, int) @ 0x000000000f8cbc23
> 2025-07-25 11:21:24 5. ./src/IO/ReadBufferFromMemory.h:36: DB::FST::FiniteStateTransducer::getOutput(std::basic_string_view<char, std::char_traits<char>>) @ 0x000000000fab927a
> 2025-07-25 11:21:24 6. ./ci/tmp/build/./src/Storages/MergeTree/GinIndexStore.cpp:674: DB::GinIndexStoreDeserializer::readSegmentedPostingsLists(String const&) @ 0x0000000018471682
> 2025-07-25 11:21:24 7. ./ci/tmp/build/./src/Storages/MergeTree/GinIndexStore.cpp:697: DB::GinIndexStoreDeserializer::createPostingsCacheFromTerms(std::vector<String, std::allocator<String>> const&) @ 0x00000000184718e2
> 2025-07-25 11:21:24 8. ./ci/tmp/build/./src/Interpreters/GinFilter.cpp:190: DB::GinFilter::contains(DB::GinFilter const&, DB::PostingsCacheForStore&, DB::GinSearchMode) const @ 0x00000000186caafd
> 2025-07-25 11:21:24 9. ./ci/tmp/build/./src/Storages/MergeTree/MergeTreeIndexGin.cpp:290: DB::MergeTreeIndexConditionGin::mayBeTrueOnGranuleInPart(std::shared_ptr<DB::IMergeTreeIndexGranule>, DB::PostingsCacheForStore&) const @ 0x00000000186bcdd3
> 2025-07-25 11:21:24 10. ./ci/tmp/build/./src/Storages/MergeTree/MergeTreeDataSelectExecutor.cpp:1802: DB::MergeTreeDataSelectExecutor::filterMarksUsingIndex(std::shared_ptr<DB::IMergeTreeIndex const>, std::shared_ptr<DB::IMergeTreeIndexCondition>, std::shared_ptr<DB::IMergeTreeDataPart const>, DB::MarkRanges const&, DB::RangesInDataPartReadHints const&, DB::Settings const&, DB::MergeTreeReaderSettings const&, DB::MarkCache*, DB::UncompressedCache*, DB::VectorSimilarityIndexCache*, std::shared_ptr<Poco::Logger>) @ 0x000000001869bf24
> 2025-07-25 11:21:24 11. ./ci/tmp/build/./src/Storages/MergeTree/MergeTreeDataSelectExecutor.cpp:807: DB::MergeTreeDataSelectExecutor::filterPartsByPrimaryKeyAndSkipIndexes(DB::RangesInDataParts, std::shared_ptr<DB::StorageInMemoryMetadata const>, std::shared_ptr<DB::Context const> const&, DB::KeyCondition const&, std::optional<DB::KeyCondition> const&, std::optional<DB::KeyCondition> const&, DB::UsefulSkipIndexes const&, DB::MergeTreeReaderSettings const&, std::shared_ptr<Poco::Logger>, unsigned long, std::vector<DB::ReadFromMergeTree::IndexStat, std::allocator<DB::ReadFromMergeTree::IndexStat>>&, bool, bool, bool)::$_1::operator()(unsigned long) const @ 0x0000000018696fce
> 2025-07-25 11:21:24 12. ./ci/tmp/build/./src/Storages/MergeTree/MergeTreeDataSelectExecutor.cpp:856: DB::MergeTreeDataSelectExecutor::filterPartsByPrimaryKeyAndSkipIndexes(DB::RangesInDataParts, std::shared_ptr<DB::StorageInMemoryMetadata const>, std::shared_ptr<DB::Context const> const&, DB::KeyCondition const&, std::optional<DB::KeyCondition> const&, std::optional<DB::KeyCondition> const&, DB::UsefulSkipIndexes const&, DB::MergeTreeReaderSettings const&, std::shared_ptr<Poco::Logger>, unsigned long, std::vector<DB::ReadFromMergeTree::IndexStat, std::allocator<DB::ReadFromMergeTree::IndexStat>>&, bool, bool, bool) @ 0x0000000018693d3b
> 2025-07-25 11:21:24 13. ./ci/tmp/build/./src/Processors/QueryPlan/ReadFromMergeTree.cpp:2017: DB::ReadFromMergeTree::selectRangesToRead(DB::RangesInDataParts, std::shared_ptr<DB::MergeTreeData::IMutationsSnapshot const>, std::optional<DB::VectorSearchParameters> const&, std::shared_ptr<DB::StorageInMemoryMetadata const> const&, DB::SelectQueryInfo const&, std::shared_ptr<DB::Context const>, unsigned long, std::shared_ptr<std::unordered_map<String, long, std::hash<String>, std::equal_to<String>, std::allocator<std::pair<String const, long>>> const>, DB::MergeTreeData const&, std::vector<String, std::allocator<String>> const&, std::shared_ptr<Poco::Logger>, std::optional<DB::ReadFromMergeTree::Indexes>&, bool) @ 0x000000001921afbc
> 2025-07-25 11:21:24 14. ./ci/tmp/build/./src/Processors/QueryPlan/ReadFromMergeTree.cpp:1687: DB::ReadFromMergeTree::selectRangesToRead(bool) const @ 0x000000001921959d
> 2025-07-25 11:21:24 15. ./ci/tmp/build/./src/Processors/QueryPlan/ReadFromMergeTree.cpp:2282: DB::ReadFromMergeTree::initializePipeline(DB::QueryPipelineBuilder&, DB::BuildQueryPipelineSettings const&) @ 0x000000001922327e
> 2025-07-25 11:21:24 16. ./ci/tmp/build/./src/Processors/QueryPlan/ISourceStep.cpp:20: DB::ISourceStep::updatePipeline(std::vector<std::unique_ptr<DB::QueryPipelineBuilder, std::default_delete<DB::QueryPipelineBuilder>>, std::allocator<std::unique_ptr<DB::QueryPipelineBuilder, std::default_delete<DB::QueryPipelineBuilder>>>>, DB::BuildQueryPipelineSettings const&) @ 0x00000000191b2e4c
> 2025-07-25 11:21:24 17. ./ci/tmp/build/./src/Processors/QueryPlan/QueryPlan.cpp:202: DB::QueryPlan::buildQueryPipeline(DB::QueryPlanOptimizationSettings const&, DB::BuildQueryPipelineSettings const&, bool) @ 0x00000000191e4144
> 2025-07-25 11:21:24 18. ./ci/tmp/build/./src/Interpreters/InterpreterSelectQueryAnalyzer.cpp:289: DB::InterpreterSelectQueryAnalyzer::buildQueryPipeline() @ 0x00000000157ab917
> 2025-07-25 11:21:24 19. ./ci/tmp/build/./src/Interpreters/InterpreterSelectQueryAnalyzer.cpp:256: DB::InterpreterSelectQueryAnalyzer::execute() @ 0x00000000157ab65c
> 2025-07-25 11:21:24 20. ./ci/tmp/build/./src/Interpreters/executeQuery.cpp:1561: DB::executeQueryImpl(char const*, char const*, std::shared_ptr<DB::Context>, DB::QueryFlags, DB::QueryProcessingStage::Enum, DB::ReadBuffer*, std::shared_ptr<DB::IAST>&, std::shared_ptr<DB::ImplicitTransactionControlExecutor>) @ 0x0000000015acd78d
> 2025-07-25 11:21:24 21. ./ci/tmp/build/./src/Interpreters/executeQuery.cpp:1769: DB::executeQuery(String const&, std::shared_ptr<DB::Context>, DB::QueryFlags, DB::QueryProcessingStage::Enum) @ 0x0000000015ac79da
> ```

interesting.. I'll have a look 👍

ahmadov committed:
fix fst for empty data
